File: shared/src/commonMain/kotlin/org/nihongo/mochi/ui/about/AboutScreen.kt
package org.nihongo.mochi.ui.about

import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.Send
import androidx.compose.material.icons.filled.Info
import androidx.compose.material.icons.filled.Star
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import org.jetbrains.compose.resources.painterResource
import org.jetbrains.compose.resources.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import org.nihongo.mochi.presentation.MochiBackground
import org.nihongo.mochi.shared.generated.resources.Res
import org.nihongo.mochi.shared.generated.resources.*

@Composable
fun AboutScreen(
    versionName: String,
    currentDate: String,
    onIssueTrackerClick: () -> Unit,
    onRateAppClick: () -> Unit,
    onPatreonClick: () -> Unit,
    onTipeeeClick: () -> Unit,
    onKanjiDataClick: () -> Unit
) {
    val scrollState = rememberScrollState()

    MochiBackground {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(scrollState)
                .padding(16.dp)
        ) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(bottom = 4.dp),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    text = stringResource(Res.string.menu_about),
                    fontSize = 34.sp,
                    fontWeight = FontWeight.Bold,
                    color = MaterialTheme.colorScheme.primary
                )
                
                Image(
                    painter = painterResource(Res.drawable.nihongomochi),
                    contentDescription = stringResource(Res.string.app_name),
                    modifier = Modifier.size(96.dp)
                )
            }

            Text(
                text = stringResource(Res.string.about_version_info),
                fontSize = 16.sp,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.padding(bottom = 24.dp)
            )

            AboutSectionCard(
                title = stringResource(Res.string.about_category_informations),
                icon = Icons.Default.Info
            ) {
                Row(modifier = Modifier.padding(bottom = 8.dp)) {
                    Text(
                        text = stringResource(Res.string.about_version_label),
                        fontWeight = FontWeight.Bold,
                        modifier = Modifier.padding(end = 16.dp),
                        color = MaterialTheme.colorScheme.onSurface
                    )
                    Text(
                        text = versionName,
                        color = MaterialTheme.colorScheme.onSurface
                    )
                }
                Row(modifier = Modifier.padding(bottom = 16.dp)) {
                    Text(
                        text = stringResource(Res.string.about_date_label),
                        fontWeight = FontWeight.Bold,
                        modifier = Modifier.padding(end = 16.dp),
                        color = MaterialTheme.colorScheme.onSurface
                    )
                    Text(
                        text = currentDate,
                        color = MaterialTheme.colorScheme.onSurface
                    )
                }

                FullWidthButton(
                    text = stringResource(Res.string.about_issue_tracker),
                    icon = Icons.AutoMirrored.Filled.Send,
                    onClick = onIssueTrackerClick
                )
                
                FullWidthButton(
                    text = stringResource(Res.string.about_rate_app),
                    icon = Icons.Default.Star, 
                    onClick = onRateAppClick
                )
            }

            Spacer(modifier = Modifier.height(16.dp))

            AboutSectionCard(
                title = stringResource(Res.string.about_category_credits),
                icon = Icons.Default.Info 
            ) {
                Text(
                    text = stringResource(Res.string.about_design_dev),
                    fontWeight = FontWeight.Bold,
                    color = MaterialTheme.colorScheme.primary,
                    modifier = Modifier.padding(bottom = 4.dp)
                )
                Text(
                    text = "LECOQ Vincent",
                    fontSize = 16.sp,
                    color = MaterialTheme.colorScheme.onSurface,
                    modifier = Modifier.padding(bottom = 8.dp)
                )

                FullWidthButton(
                    text = stringResource(Res.string.about_patreon),
                    onClick = onPatreonClick
                )
                
                FullWidthButton(
                    text = stringResource(Res.string.about_tipeee),
                    onClick = onTipeeeClick,
                    modifier = Modifier.padding(bottom = 16.dp)
                )

                Text(
                    text = stringResource(Res.string.about_pedagogical),
                    fontWeight = FontWeight.Bold,
                    color = MaterialTheme.colorScheme.primary,
                    modifier = Modifier.padding(bottom = 4.dp)
                )
                Text(
                    text = stringResource(Res.string.about_coming_soon),
                    fontSize = 16.sp,
                    color = MaterialTheme.colorScheme.onSurface,
                    modifier = Modifier.padding(bottom = 8.dp)
                )
            }
//Deprecated
//            Spacer(modifier = Modifier.height(16.dp))
//
//            AboutSectionCard(
//                title = stringResource(Res.string.about_category_resources),
//                icon = Icons.Default.Info
//            ) {
//                 FullWidthButton(
//                    text = stringResource(Res.string.about_kanji_data_credit),
//                    onClick = onKanjiDataClick
//                )
//            }
//
            Spacer(modifier = Modifier.height(16.dp))
        }
    }
}

@Composable
fun AboutSectionCard(
    title: String,
    icon: ImageVector,
    content: @Composable () -> Unit
) {
    Card(
        modifier = Modifier.fillMaxWidth(),
        shape = RoundedCornerShape(16.dp),
        elevation = CardDefaults.cardElevation(2.dp),
        colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surface.copy(alpha = 0.9f))
    ) {
        Column(modifier = Modifier.padding(16.dp)) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(bottom = 16.dp),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    text = title,
                    fontSize = 20.sp,
                    fontWeight = FontWeight.Bold,
                    color = MaterialTheme.colorScheme.onSurface
                )
                Icon(
                    imageVector = icon,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.primary
                )
            }
            
            content()
        }
    }
}

@Composable
fun FullWidthButton(
    text: String,
    icon: ImageVector? = null,
    onClick: () -> Unit,
    modifier: Modifier = Modifier
) {
    Button(
        onClick = onClick,
        modifier = modifier.fillMaxWidth(),
        colors = ButtonDefaults.buttonColors(
            containerColor = MaterialTheme.colorScheme.primary,
            contentColor = MaterialTheme.colorScheme.onPrimary
        ),
        elevation = ButtonDefaults.buttonElevation(defaultElevation = 6.dp),
        shape = RoundedCornerShape(4.dp)
    ) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.Center,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(text = text)
            if (icon != null) {
                Spacer(modifier = Modifier.width(8.dp))
                Icon(
                    imageVector = icon, 
                    contentDescription = null,
                    modifier = Modifier.size(18.dp)
                )
            }
        }
    }
}

File: flutter_app/lib/about_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:url_launcher/url_launcher.dart';
import 'package:package_info_plus/package_info_plus.dart';
import 'build_info.g.dart';
import 'widgets/mochi_background.dart';
import 'providers/settings_provider.dart';

class AboutScreen extends StatefulWidget {
  const AboutScreen({super.key});

  @override
  State<AboutScreen> createState() => _AboutScreenState();
}

class _AboutScreenState extends State<AboutScreen> {
  String _version = "0.9.8";
  String _buildDate = "";

  @override
  void initState() {
    super.initState();
    _loadPackageInfo();
  }

  Future<void> _loadPackageInfo() async {
    final info = await PackageInfo.fromPlatform();
    
    final year = kBuildTimestamp.year;
    final month = kBuildTimestamp.month.toString().padLeft(2, '0');
    final day = kBuildTimestamp.day.toString().padLeft(2, '0');

    if (mounted) {
      setState(() {
        _version = info.version;
        _buildDate = "$year-$month-$day";
      });
    }
  }

  Future<void> _launchUrl(String url) async {
    final Uri uri = Uri.parse(url);
    if (!await launchUrl(uri, mode: LaunchMode.externalApplication)) {
      if (mounted) {
        ScaffoldMessenger.of(context).showSnackBar(
          SnackBar(content: Text('Impossible d\'ouvrir : $url')),
        );
      }

    }
  }

  @override
  Widget build(BuildContext context) {
    final settings = context.watch<SettingsProvider>();
    final theme = Theme.of(context);

    return Scaffold(
      appBar: AppBar(
        title: Text(
          settings.getString("menu_about"),
          style: TextStyle(color: theme.colorScheme.onSurface),
        ),
        backgroundColor: Colors.transparent,
        elevation: 0,
        iconTheme: IconThemeData(color: theme.colorScheme.onSurface),
      ),
      extendBodyBehindAppBar: true,
      body: MochiBackground(
        child: SafeArea(
          child: SingleChildScrollView(
            padding: const EdgeInsets.all(16.0),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: [
                    Text(
                      "Nihongo\nMochi",
                      style: TextStyle(
                        fontSize: 34,
                        fontWeight: FontWeight.bold,
                        color: theme.colorScheme.primary,
                      ),
                    ),
                    Image.asset('assets/drawable/nihongomochi.webp', width: 96),
                  ],
                ),
                const SizedBox(height: 24),
                _buildSectionCard(
                  context,
                  title: settings.getString("about_category_informations"),
                  icon: Icons.info_outline,
                  children: [
                    _buildInfoRow(context, settings.getString("about_version_label"), _version),
                    _buildInfoRow(context, settings.getString("about_date_label"), _buildDate),
                    const SizedBox(height: 16),
                    Row(
                      children: [
                        ClipRRect(
                          borderRadius: BorderRadius.circular(8),
                          child: Image.asset('assets/drawable/ebi.webp', width: 48, height: 48, errorBuilder: (_, _, _) => const Icon(Icons.pets, size: 48)),

                        ),
                        const SizedBox(width: 12),
                        Expanded(
                          child: Text(
                            settings.getString("in_memoriam"),
                            style: TextStyle(
                              fontSize: 12,
                              color: theme.colorScheme.onSurface.withValues(alpha: 0.7),
                              fontStyle: FontStyle.italic,
                            ),
                          ),
                        ),
                      ],
                    ),
                    const SizedBox(height: 16),
                    _buildFullWidthButton(context, settings.getString("about_issue_tracker"), Icons.bug_report, () => _launchUrl("https://github.com/oktailb/NihongoMochi/issues")),
                    _buildFullWidthButton(context, settings.getString("about_rate_app"), Icons.star, () {}),
                    _buildFullWidthButton(context, "Open Source Licenses", Icons.description, () => showLicensePage(context: context)),
                  ],
                ),
                const SizedBox(height: 16),
                _buildSectionCard(
                  context,
                  title: settings.getString("about_category_credits"),
                  icon: Icons.people_outline,
                  children: [
                    Text(
                      settings.getString("about_design_dev"),
                      style: TextStyle(fontWeight: FontWeight.bold, color: theme.colorScheme.primary),
                    ),
                    Text(
                      settings.getString("about_author_name"),
                      style: TextStyle(color: theme.colorScheme.onSurface),
                    ),
                    const SizedBox(height: 16),
                    _buildFullWidthButton(context, settings.getString("about_patreon"), Icons.favorite, () => _launchUrl("https://www.patreon.com/cw/Oktail")),
                    _buildFullWidthButton(context, settings.getString("about_tipeee"), Icons.volunteer_activism, () => _launchUrl("https://en.tipeee.com/lecoq-vincent/news/246869")),
                    const SizedBox(height: 16),
                    Text(
                      settings.getString("about_pedagogical"),
                      style: TextStyle(fontWeight: FontWeight.bold, color: theme.colorScheme.primary),
                    ),
                    Text(
                      settings.getString("about_coming_soon"),
                      style: TextStyle(color: theme.colorScheme.onSurface),
                    ),
                  ],
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }

  Widget _buildSectionCard(BuildContext context, {required String title, required IconData icon, required List<Widget> children}) {
    final theme = Theme.of(context);
    return Card(
      shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(16)),
      color: theme.colorScheme.surface.withValues(alpha: 0.9),
      elevation: 2,
      child: Padding(
        padding: const EdgeInsets.all(16.0),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                Text(
                  title,
                  style: TextStyle(fontSize: 20, fontWeight: FontWeight.bold, color: theme.colorScheme.onSurface),
                ),
                Icon(icon, color: theme.colorScheme.primary),
              ],
            ),
            const SizedBox(height: 16),
            ...children,
          ],
        ),
      ),
    );
  }

  Widget _buildInfoRow(BuildContext context, String label, String value) {
    final theme = Theme.of(context);
    return Padding(
      padding: const EdgeInsets.only(bottom: 8.0),
      child: Row(
        children: [
          Text(
            label,
            style: TextStyle(fontWeight: FontWeight.bold, color: theme.colorScheme.onSurface),
          ),
          const SizedBox(width: 16),
          Text(
            value,
            style: TextStyle(color: theme.colorScheme.onSurface),
          ),
        ],
      ),
    );
  }

  Widget _buildFullWidthButton(BuildContext context, String text, IconData icon, VoidCallback onTap) {
    final theme = Theme.of(context);
    return Padding(
      padding: const EdgeInsets.only(bottom: 8.0),
      child: ElevatedButton.icon(
        onPressed: onTap,
        icon: Icon(icon, size: 18),
        label: Text(text),
        style: ElevatedButton.styleFrom(
          minimumSize: const Size(double.infinity, 45),
          backgroundColor: theme.colorScheme.primary,
          foregroundColor: theme.colorScheme.onPrimary,
          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(8)),
          elevation: 4,
        ),
      ),
    );
  }
}
